### ISSUE TITLE
backpressure & recovery

### ISSUE THREAD

**Issue by edsiper** [fluent/fluent-bit#3341]
## Bug Report

[internal] on specific backpressure and disk saturation conditions, the service delays logs delivery.

**Comment by edsiper** [comment]
[cc: @hsmatulisgoogle @qingling123]

### Troubleshooting

- storage backlog size: 25G
- storage backlog chunks: 12k
- remote endpoint: down

### Storage Reported Stats

```
===== Storage Layer =====
total chunks     : 12692
├─ mem chunks    : 0
└─ fs chunks     : 12692
   ├─ up         : 100
   └─ down       : 12592
[2021/04/07 23:41:30] [engine] caught signal (SIGCONT)
[2021/04/07 23:41:30] Fluent Bit Dump

===== Input =====
tail.0 (tail)
│
├─ status
│  └─ overlimit     : no
│     ├─ mem size   : 197.1M (206690020 bytes)
│     └─ mem limit  : 0b (0 bytes)
│
├─ tasks
│  ├─ total tasks   : 0
│  ├─ new           : 0
│  ├─ running       : 0
│  └─ size          : 0b (0 bytes)
│
└─ chunks
   └─ total chunks  : 125
      ├─ up chunks  : 1
      ├─ down chunks: 124
      └─ busy chunks: 0
         ├─ size    : 0b (0 bytes)
         └─ size err: 0

storage_backlog.1 (storage_backlog)
│
├─ status
│  └─ overlimit     : no
│     ├─ mem size   : 0b (0 bytes)
│     └─ mem limit  : 0b (0 bytes)
│
├─ tasks
│  ├─ total tasks   : 100
│  ├─ new           : 0
│  ├─ running       : 100
│  └─ size          : 197.1M (206690020 bytes)
│
└─ chunks
   └─ total chunks  : 100
      ├─ up chunks  : 100
      ├─ down chunks: 0
      └─ busy chunks: 100
         ├─ size    : 197.1M (206690020 bytes)
         └─ size err: 0


===== Storage Layer =====
total chunks     : 12719
├─ mem chunks    : 0
└─ fs chunks     : 12719
   ├─ up         : 100
   └─ down       : 12619
```

## Findings

- Since the remote endpoint is down, chunks that are ```up``` are in the retry list, each chunk might be retried in X number of seconds, but the chunk is limiting new chunks to be up in memory.

- New ingestion of data get stored properly, but cannot be flushed even if the remote endpoint is UP because we are hitting the ```storage.max_chunks_up``` limit (100)

- If Fluent Bit starts with storage backlog data, the backlog chunks will get priority using all the limits for the allowed chunks up. So no data will be delivered until chunks that are up in memory gets processed

## Possible solutions

- Limit the number of chunks that storage can hold up to avoid consuming the whole capacity. 

- Allow max_chunks_up per input plugin to make sure it can try to process data on hard backpressure conditions

**Comment by qingling128** [comment]
Hi @edsiper  any luck with this investigation?

**Comment by rmacian** [comment]
may be I am hitting this. After the receiver recovers from crash, I get lots of retrys and connection refused despite the receiver is up. If I restart the container that is retrying I get lots of corrupted events like
`[2021/04/19 15:31:37] [error] [input:storage_backlog:storage_backlog.9] chunk validation failed, data might be corrupted. Found 1050 valid records, failed content starts right after byte 1413117.
[2021/04/19 15:31:37] [error] [input:storage_backlog:storage_backlog.9] removing chunk tail.0:1-1618842192.722514440.flb from the queue
[2021/04/19 15:31:37] [error] [input:storage_backlog:storage_backlog.9] chunk validation failed, data might be corrupted. Found 1126 valid records, failed content starts right after byte 1525032`

**Comment by pranavmarla** [comment]
I believe this is similar to issue [#3040](https://github.com/fluent/fluentd/issues/3040) from the Fluentd repo -- any idea if a similar fix is in the works there?

**Comment by debu99** [comment]
i am hitting the same issue

**Comment by qingling128** [comment]
Some mitigations are implemented in:
* https://github.com/fluent/fluent-bit/pull/3498 (network: asynchronous DNS support)
* https://github.com/fluent/fluent-bit/pull/3498 (network: asynchronous DNS support)

They are both released in 1.7.6.

Previously the issues we saw that triggered this are due to blocking DNS lookup and deadlock in oauth token refresh.
